accounts/abi/bind/dep_tree_test.go

```diff
@@ -0,0 +1,309 @@
+}
+
+func testLinkCase(t *testing.T, tcInput linkTestCaseInput) {
+	testAddr := crypto.PubkeyToAddress(testKey.PublicKey)
+	var testAddrNonce uint64
+	overridesAddrs := make(map[common.Address]struct{})
+
+	// generate deterministic addresses for the override set.
+	rand.Seed(42)
+	overrideAddrs := make(map[rune]common.Address)
+	for contract, _ := range tcInput.overrides {
+		var addr common.Address
+		rand.Read(addr[:])
+		overrideAddrs[contract] = addr
+		overridesAddrs[addr] = struct{}{}
+	}
+
+	tc := makeLinkTestCase(tcInput.input, overrideAddrs)
+	allContracts := make(map[rune]struct{})
+
+	for contract, deps := range tcInput.input {
+		allContracts[contract] = struct{}{}
+		for _, dep := range deps {
+			allContracts[dep] = struct{}{}
+		}
+	}
+
+	mockDeploy := func(input []byte, deployer []byte) (common.Address, *types.Transaction, error) {
+		contractAddr := crypto.CreateAddress(testAddr, testAddrNonce)
+		testAddrNonce++
+
+		if len(deployer) >= 20 {
+			// assert that this contract only references libs that are known to be deployed or in the override set
+			for i := 0; i < len(deployer); i += 20 {
+				var dep common.Address
+				dep.SetBytes(deployer[i : i+20])
+				if _, ok := overridesAddrs[dep]; !ok {
+					t.Fatalf("reference to dependent contract that has not yet been deployed: %x\n", dep)
+				}
+			}
+		}
+		overridesAddrs[contractAddr] = struct{}{}
+		// we don't care about the txs themselves for the sake of the linking tests.  so we can return nil for them in the mock deployer
+		return contractAddr, nil, nil
+	}
+
+	contracts := make(map[string]*MetaData)
+	overrides := make(map[string]common.Address)
+
+	for pattern, bin := range tc.contractCodes {
+		contracts[pattern] = &MetaData{Pattern: pattern, Bin: "0x" + bin}
+	}
+	for pattern, bin := range tc.libCodes {
+		contracts[pattern] = &MetaData{
+			Bin:     "0x" + bin,
+			Pattern: pattern,
+		}
+	}
+
+	contractsList := linkDeps(contracts)
+
+	for pattern, override := range tc.overrides {
+		overrides[pattern] = override
+	}
+
+	deployParams := NewDeploymentParams(contractsList, nil, overrides)
+	res, err := LinkAndDeploy(deployParams, mockDeploy)
+	if err != nil {
+		t.Fatalf("got error from LinkAndDeploy: %v\n", err)
+	}
+
+	if len(res.Addrs) != len(tcInput.expectDeployed) {
+		t.Fatalf("got %d deployed contracts.  expected %d.\n", len(res.Addrs), len(tcInput.expectDeployed))
+	}
+	for contract, _ := range tcInput.expectDeployed {
+		pattern := crypto.Keccak256Hash([]byte(string(contract))).String()[2:36]
+		if _, ok := res.Addrs[pattern]; !ok {
+			t.Fatalf("expected contract %s was not deployed\n", string(contract))
+		}
+	}
+}
+
+func TestContractLinking(t *testing.T) {
+	// test simple contract without any dependencies or overrides
+	testLinkCase(t, linkTestCaseInput{
+		map[rune][]rune{
+			'a': {}},
+		map[rune]struct{}{},
+		map[rune]struct{}{
+			'a': {}}})
+	// test deployment of a contract that depends on somes libraries.
+	testLinkCase(t, linkTestCaseInput{
```

Comment:
This is a bit clunky. When these fail, you typicall get
```
[user@work go-ethereum]$ go test ./accounts/abi/bind -run TestContractLinking
--- FAIL: TestContractLinking (0.00s)
    dep_tree_test.go:197: got 5 deployed contracts.  expected 1.

```
If you make it a proper table-driven test, it can spit out also _which_ test it was that failed. So convert this to a slice of `linkTestCaseinput`, and iterate over that slice